pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -218,15 +256,73 @@ pub mod pallet {
         UnexpectedXCMResponse,
         /// Block production is pending for para id with successfully placed order
         BlockProductionPending,
+        /// Block production is not allowed for current slot
+        NotAllowedToProduceBlockRightNow,
+        /// Collator signature nonce is incorrect
+        IncorrectCollatorSignatureNonce,
+        /// Collator signature is invalid
+        InvalidCollatorSignature,
+    }
+
+    impl<T: Config> From<BuyingError<BlockNumberFor<T>>> for Error<T> {
+        fn from(value: BuyingError<BlockNumberFor<T>>) -> Self {
+            match value {
+                BuyingError::OrderAlreadyExists { .. } => Error::<T>::OrderAlreadyExists,
+                BuyingError::BlockProductionPending { .. } => Error::<T>::BlockProductionPending,
+                BuyingError::NotAParathread => Error::<T>::NotAParathread,
+                BuyingError::NotAllowedToProduceBlockRightNow { .. } => {
+                    Error::<T>::NotAllowedToProduceBlockRightNow
+                }
+            }
+        }
     }
 
     /// Proof that I am a collator, assigned to a para_id, and I can buy a core for that para_id
     #[derive(Encode, Decode, CloneNoBound, PartialEq, Eq, DebugNoBound, TypeInfo)]
-    #[scale_info(skip_type_params(T))]
-    pub struct BuyCoreCollatorProof<T: Config> {
-        account: T::AccountId,
-        // TODO
-        _signature: (),
+    pub struct BuyCoreCollatorProof<PublicKey>
+    where
+        PublicKey: RuntimeAppPublic + Clone + core::fmt::Debug,
+    {
+        nonce: u64,
+        public_key: PublicKey,
+        signature: PublicKey::Signature,
+    }
+
+    impl<PublicKey> BuyCoreCollatorProof<PublicKey>
+    where
+        PublicKey: RuntimeAppPublic + Clone + core::fmt::Debug,
+    {
+        fn verify_signature(&self, para_id: ParaId) -> bool {
+            let payload = (self.nonce, para_id).encode();
+            self.public_key.verify(&payload, &self.signature)
+        }
+
+        #[cfg(any(test, feature = "runtime-benchmarks"))]
+        pub fn new(nonce: u64, para_id: ParaId, public_key: PublicKey) -> Option<Self> {
```

Comment:
```suggestion
        pub fn new(nonce: u64, para_id: ParaId, public_key: PublicKey) -> Option<Self> {
```

Collators will probably need this to be public to create the proof, right?

Reply:
this should use the private key to sign, not the public one

Reply:
@girazoki It uses private key from the keystore.

Reply:
Then the generic name is wrong? It should be `private_key: PrivateKey` I guess?

Reply:
No. This is the public key. The `KeyStoreExt` underneath the `verify` call, looks up the private key from keystone and signs with it.